tests/python/unittest/test_optimizer.py

```diff
@@ -0,0 +1,45 @@
+import nnvm
+import nnvm.compiler.optimizer as optimizer
+import nnvm.compiler.lr_scheduler as lr_scheduler
+
+def test_FactorScheduler():
+    scheduler = lr_scheduler.FactorScheduler(step=5, factor=0.1)
+    num_updates = nnvm.sym.Variable('update_t')
+    update_t = nnvm.sym.assign(num_updates, num_updates + 1)
+    update_lr = scheduler(update_t)
+    g = nnvm.graph.create(update_lr).ir()
+    assert '__rpow_scalar__' in g
+    assert 'clip' in g
+
+def test_sgd():
+    x = nnvm.sym.Variable("x")
+    y = nnvm.sym.Variable("y")
+    z = nnvm.sym.elemwise_add(x, nnvm.sym.sqrt(y))
+
+    scheduler = lr_scheduler.FactorScheduler(step=10, factor=0.5)
+    opt = optimizer.SGD(learning_rate = 0.001, lr_scheduler=scheduler,
+                        rescale_grad=0.2, clip_gradient=1.0, wd=0.1)
+    opt_sym = opt.minimize(z)
+    assert len(opt_sym.list_output_names()) == 2
+    g = nnvm.graph.create(opt_sym).ir()
+    assert 'SGD_t' in g
+
+def test_adam():
+    x = nnvm.sym.Variable('x')
+    y = nnvm.sym.elemwise_add(x, x, name='add1')
+    y = nnvm.sym.flatten(y, name="flatten1")
+
+    scheduler = lr_scheduler.FactorScheduler(step=5, factor=0.6)
+    opt = optimizer.Adam(lr_scheduler=scheduler,
+                         rescale_grad=0.2, clip_gradient=1.0, wd=0.2)
+    opt_sym = opt.minimize(y)
+    assert len(opt_sym.list_output_names()) == 1
+    g = nnvm.graph.create(opt_sym).ir()
+    assert 'Adam_t' in g
+    assert 'Adam_m0' in g
+    assert 'Adam_v0' in g
+
```

Comment:
please add a test-case for stateful ops with simple assign. We might need to update graph runtime to add get_input, as the states belongs to input variable

Reply:
Yes. We need to consider such kind of in-place manner operators. Currently it looks like in tvm there is no such function primitives.

Reply:
In tvm, assign can be prototyped as writing tensor values directly to assigned tensor memory. Do we have a elegant way to implement this in tvm?

Reply:
Or in short term, I can go ahead to address comments and make nnvm training graph complete. We can consider tvm implementation later.

Reply:
We can pass in arguments for both input and output in TVM, for elemwise operation they can be the same.

For example
```python
x = tvm.placeholder(shape)
y = tvm.placeholder(shape)
z = topi.elemwise_add(x, y)

s = tvm.create_schedule(z.op)
f = tvm.build(s, [x, y, z])

data = tvm.nd.array(shape)
rhs = tvm.nd.array(shape)
# data = data + rhs
f(data, rhs, data)
```

Reply:
While we can have a built function for assign in tvm, we still need to register in nnvm. However, in nnvm register_compute function, we can't build a tvm function and utilize tvm.nd.array interface to move tensor values. Is my understanding correct?

Reply:
That is true, a special registration mechanism is needed to tell NNVM compiler to generate the operator that mutates its input. Need to update the NNVM compiler's build logic to support this

Reply:
Is it possible to directly manipulate tvm tensor internal data in FTVMCompute function? 
```
.set_attr<FTVMCompute>(
  "FTVMCompute", [](const NodeAttrs& attrs,
                    const Array<Tensor>& inputs,
                    const Array<Tensor>& out_info) {
      rtv = Array<Tensor>{ topi::identity(inputs[1]) };
      // Some manipulation to change internal data structure to make inputs[0] 'points' to rtv
      inputs[0]...
      return rtv;
})
```